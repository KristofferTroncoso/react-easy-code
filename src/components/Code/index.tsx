export {Code} from './Code';
export {InlineCode} from './InlineCode';
export {BlockCode} from './BlockCode';
