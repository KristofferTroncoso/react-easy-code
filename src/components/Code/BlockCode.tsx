import React, { useEffect } from "react";
import Prism from 'prismjs';
import 'prismjs/plugins/normalize-whitespace/prism-normalize-whitespace';
import 'prismjs/components/prism-jsx';
import styled from 'styled-components';

const StyledPre = styled.pre`
  color: ${props => props.theme === "light" ? 'black ': 'white'};
	background: ${props => props.theme === "light" ? '#f5f2f0': '#333'};
	text-shadow: ${props => props.theme === "light" ? '0 1px white': 'none'};
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  font-size: 1em;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;
  tab-size: 4;
  hyphens: none;
  padding: 1em;
  margin: .5em 0;
  overflow: auto;

  ::selection {
    text-shadow: none;
    background: #b3d4fc;
  }

  @media print {
      text-shadow: none;
  }

  .token.comment,
  .token.prolog,
  .token.doctype,
  .token.cdata {
    color: slategray;
  }

  .token.punctuation {
    color: #999;
  }

  .namespace {
    opacity: .7;
  }

  .token.property,
  .token.tag,
  .token.boolean,
  .token.number,
  .token.constant,
  .token.symbol,
  .token.deleted {
    color: ${props => props.theme === "light" ? '#905': '#c43181'};
  }

  .token.selector,
  .token.attr-name,
  .token.string,
  .token.char,
  .token.builtin,
  .token.inserted {
    color: #690;
  }

  .token.operator,
  .token.entity,
  .token.url,
  .language-css .token.string,
  .style .token.string {
    color: #9a6e3a;
  }

  .token.atrule,
  .token.attr-value,
  .token.keyword {
    color: ${props => props.theme === "light" ? '#07a': '#00a0e5'};
  }

  .token.function,
  .token.class-name {
    color: #DD4A68;
  }

  .token.regex,
  .token.important,
  .token.variable {
    color: #e90;
  }

  .token.important,
  .token.bold {
    font-weight: bold;
  }
  .token.italic {
    font-style: italic;
  }

  .token.entity {
    cursor: help;
  }
`;

export interface BlockCodeProps {
  render: string
  lang?: "jsx" | "css" | "javascript"
  theme?: "light" | "dark"
}

export function BlockCode({lang = "jsx", render, theme="light" }: BlockCodeProps) {
  let strLang = `language-${lang}`;

  useEffect(() => {
    Prism.highlightAll();
  }, []);

  return (
    <StyledPre className={strLang} theme={theme}>
      <code>{render}</code>
    </StyledPre>
  );
}