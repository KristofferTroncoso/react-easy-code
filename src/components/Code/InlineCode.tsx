import React, { useEffect } from "react";
import Prism from 'prismjs';
import 'prismjs/plugins/normalize-whitespace/prism-normalize-whitespace';
import 'prismjs/components/prism-jsx';
import styled from 'styled-components';

const StyledCode = styled.code`
	color: ${props => props.theme === "light" ? 'black ': 'white'};
	background: ${props => props.theme === "light" ? '#f5f2f0': '#333'};
	text-shadow: ${props => props.theme === "light" ? '0 1px white': 'none'};
	font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
	font-size: 1em;
	text-align: left;
	word-spacing: normal;
	word-break: normal;
	word-wrap: normal;
	line-height: 1.5;
	tab-size: 4;
	hyphens: none;
  padding: .1em;
  border-radius: .3em;
  white-space: normal;

  ::selection {
    text-shadow: none;
    background: #b3d4fc;
  }

  @media print {
    text-shadow: none;
  }


  .token.comment,
  .token.prolog,
  .token.doctype,
  .token.cdata {
    color: slategray;
  }

  .token.punctuation {
    color: #999;
  }

  .namespace {
    opacity: .7;
  }

  .token.property,
  .token.tag,
  .token.boolean,
  .token.number,
  .token.constant,
  .token.symbol,
  .token.deleted {
    color: ${props => props.theme === "light" ? '#905': '#c43181'};
  }

  .token.selector,
  .token.attr-name,
  .token.string,
  .token.char,
  .token.builtin,
  .token.inserted {
    color: #690;
  }

  .token.operator,
  .token.entity,
  .token.url,
  .language-css .token.string,
  .style .token.string {
    color: #9a6e3a;
  }

  .token.atrule,
  .token.attr-value,
  .token.keyword {
    color: ${props => props.theme === "light" ? '#07a': '#00a0e5'};
  }

  .token.function,
  .token.class-name {
    color: #DD4A68;
  }

  .token.regex,
  .token.important,
  .token.variable {
    color: #e90;
  }

  .token.important,
  .token.bold {
    font-weight: bold;
  }
  .token.italic {
    font-style: italic;
  }

  .token.entity {
    cursor: help;
  }

`;

export interface InlineCodeProps {
  render: string
  lang?: "jsx" | "css" | "javascript"
  theme?: "light" | "dark"
}

export function InlineCode({lang = "jsx", render, theme="light"}: InlineCodeProps) {
  let strLang = `language-${lang}`;

  useEffect(() => {
    Prism.highlightAll();
  }, []);

  return (
    <StyledCode className={strLang} theme={theme}>{render}</StyledCode>
  );
}
