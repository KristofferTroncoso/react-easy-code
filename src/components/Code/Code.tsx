import React from "react";
import { InlineCode } from "./InlineCode";
import { BlockCode } from "./BlockCode";
import PropTypes from 'prop-types';

export interface CodeProps {
  render: string
  lang?: "jsx" | "css" | "javascript"
  inline?: boolean
  theme?: "light" | "dark"
}

const propTypes: object = {
  render: PropTypes.string.isRequired,
  lang: PropTypes.oneOf(["jsx", "css", "javascript"]),
  inline: PropTypes.bool,
  theme: PropTypes.oneOf(["light", "dark"])
};

/**
 * @param {string} render string of code to render
 * @param {string} lang "jsx" | "css" | "javascript" - Default "jsx"
 * @param {boolean} inline boolean - Default false
 * @param {string} theme "light" | "dark" - Default "light"
 */
export const Code: React.FC<CodeProps> = (props) => {
  const { render, lang="jsx", inline=false, theme="light" } = props;
  return (
    inline
    ? <InlineCode lang={lang} render={render} theme={theme} />
    : <BlockCode lang={lang} render={render} theme={theme} />
  );
}

Code.propTypes = propTypes;
