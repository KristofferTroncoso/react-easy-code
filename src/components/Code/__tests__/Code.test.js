import React from 'react';
import {Code} from '../Code';
import Renderer from 'react-test-renderer';

test('normal render', () => {
  const wrapper = Renderer.create(<Code render={`let i = 3`} />);
  expect(wrapper).toMatchSnapshot();
});

test('inline render', () => {
  const wrapper = Renderer.create(<Code inline render={`let i = 3`} />);
  expect(wrapper).toMatchSnapshot();
});

test('dark theme', () => {
  const wrapper = Renderer.create(<Code theme="dark" render={`let i = 3`} />);
  expect(wrapper).toMatchSnapshot();
});