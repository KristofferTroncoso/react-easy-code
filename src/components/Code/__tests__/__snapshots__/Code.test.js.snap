// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dark theme 1`] = `
<pre
  className="language-jsx sc-bwzfXH dbofpk"
>
  <code>
    let i = 3
  </code>
</pre>
`;

exports[`inline render 1`] = `
<code
  className="language-jsx sc-bdVaJa jwhKeD"
>
  let i = 3
</code>
`;

exports[`normal render 1`] = `
<pre
  className="language-jsx sc-bwzfXH eHSmvX"
>
  <code>
    let i = 3
  </code>
</pre>
`;
